### ISSUE TITLE
Payload size

### ISSUE THREAD

**Issue by xogeny** [nats-io/nats#63]
The `gnatsd` enforces a maximum payload size.  In looking at the `gnatsd` debugging output, I can see that a message I am publishing exceeds this limit.  But the call to `PublishMsg` I'm making does not return an error.

Am I correct that `nats` doesn't detect when the server rejects a message (at least based on side)?  If so, is there some reason that the client doesn't/can't check for this?

**Comment by derekcollison** [comment]
The server enforces, but will send the error in an async fashion. The client should register this from the info and enforce it at the client API level. Will make sure to add it in.. Thanks..

**Comment by xogeny** [comment]
I've got a few more details that you might find useful/interesting.  I do development on a 64-bit OSX machine.  I never saw this issue.  When I cross-compiled to 64-bit Windows on OSX, I never saw this issue.  As soon as I did a build for 32-bit Windows, suddenly I saw the issue.

I went back and checked on my Mac.  When I publish a message on my Mac, I can clearly see the size of the message and it is `1467349`.  This is larger than `MAX_PAYLOAD_SIZE`.  **Yet**, `gnatsd` isn't rejecting the message.  The enforcement of the payload size only seems to be showing up on 32-bit machines?!?  Now, I have not ruled out some kind of version skew in `gnatsd`.  But I **am** quite sure that the version of `gnatsd` installed on my machine does define the `MAX_PAYLOAD_SIZE` to be 1Mb.

I have not dug through the code, so take all this with a grain of salt I suppose.  Ideally, I'd like to create a test case that demonstrates this issue clearly.  But right now I'm afraid I'm just swamped.  Until then, take all this as anecdotal.

I hope that helps in some way.

**Comment by derekcollison** [comment]
gnatsd will report the maximum_payload, you can check by telnetting to it
and trying a pub with large payload..

> telnet demo.nats.io 4222

Trying 107.170.221.32...

Connected to demo.nats.io.

Escape character is '^]'.

INFO
{"server_id":"1ec445b504f4edfb4cf7927c707dd717","version":"0.6.6","go":"go1.4.2","host":"0.0.0.0","port":4222,"auth_required":false,"ssl_required":false,"max_payload":
*1048576*}

pub foo 2000000

-ERR 'Maximum Payload Violation'

Connection closed by foreign host.

**Comment by xogeny** [comment]
OK, I did some experimenting and I figured out what was going on with the `gnatsd` diagnostics.  I was correct that under Windows I was getting an error about maximum payload size exceeded but I didn't see the issue on my Mac.  However, I was also correct that this was an issue with version skew.  A more accurate way to describe this was that I saw the error using `gnatsd` version 0.6.6 (what I was running on Windows) but not with `gnatsd` version 0.6.0 (what I was running on my Mac).

However, my original point still stands.  With both 0.6.6 and 0.6.0, that `nats` client doesn't report any error when publishing a message that is too large.

Here is my test program that tries to send a message that is twice as large as is supported.

Here is [a Gist](https://gist.github.com/xogeny/ad2e563637d868b9c54d) that demonstrates the issue.

If I run that program under Windows, I get this output:

```
2015/08/27 09:32:43 Created a message of length: 2097152
2015/08/27 09:32:43 err = <nil>
2015/08/27 09:32:43 closed = false
2015/08/27 09:32:43 LastError = <nil>
2015/08/27 09:32:43 Created a message of length: 100
2015/08/27 09:32:43 err = <nil>
```

(note, no indication of failure)

Strangely, if I run it under OSX, I get this:

```
2015/08/27 09:34:03 Created a message of length: 2097152
2015/08/27 09:34:03 err = write tcp 127.0.0.1:4222: protocol wrong type for socket
2015/08/27 09:34:03 closed = false
2015/08/27 09:34:03 LastError = write tcp 127.0.0.1:4222: protocol wrong type for socket
2015/08/27 09:34:03 Created a message of length: 100
2015/08/27 09:34:03 err = write tcp 127.0.0.1:4222: protocol wrong type for socket
```

Here it indicates an error.  In both cases, `gnats -D` reports:

```
[70479] 2015/08/27 09:33:59.447518 [INF] Starting gnatsd version 0.6.6
[70479] 2015/08/27 09:33:59.447688 [INF] Listening for client connections on 0.0.0.0:4222
[70479] 2015/08/27 09:33:59.447868 [INF] gnatsd is ready
[70479] 2015/08/27 09:34:03.324362 [DBG] 127.0.0.1:55166 - cid:1 - Client connection created
[70479] 2015/08/27 09:34:03.343344 [ERR] 127.0.0.1:55166 - cid:1 - Maximum Payload Exceeded: 2097152 vs 1048576
[70479] 2015/08/27 09:34:03.343369 [DBG] 127.0.0.1:55166 - cid:1 - Client connection closed
```

All this is with Go 1.4.2 on both platforms.

**Comment by wallyqs** [comment]
Hi, as Derek mentioned the error is reported asynchronously so in this particular case it would be useful for you to pass a DisconnectedCB for the connection since the server would disconnect you once it identifies a transgression in the max payload size.  Showing its usage in the by extending your example a little bit:

```go
package main

import (
        "log"

        gs "github.com/apcera/gnatsd/server"
        "github.com/nats-io/nats"
)

func msg(n int) []byte {
        ret := []byte{}
        for i := 0; i < n; i++ {
                ret = append(ret, byte(i))
        }
        log.Printf("Created a message of length: %d", len(ret))
        return ret
}

func init(){
        log.SetFlags(log.LstdFlags | log.Lmicroseconds)
}

func main() {

        opts := nats.DefaultOptions
        opts.Servers = []string{ nats.DefaultURL }

        nc, err := opts.Connect()
        if err != nil {
                log.Printf("Unable to connect: %v", err)
                return
        }

        disconnectedCh := make(chan struct{})
        nc.Opts.DisconnectedCB = func(conn *nats.Conn) {
                log.Printf("Disconnected!!!")
                disconnectedCh <- struct{}{}
        }

        subj := "Request"
        reply := "Reply"

        badMsg := nats.Msg{Subject: subj, Reply: reply, Data: msg(gs.MAX_PAYLOAD_SIZE * 2)}
        err = nc.PublishMsg(&badMsg)
        log.Printf("err         = %v\n", err)
        log.Printf("closed?     = %v\n", nc.IsClosed())
        log.Printf("LastError?  = %v\n", nc.LastError())
        <-disconnectedCh
        log.Printf("closed??    = %v\n", nc.IsClosed())
        log.Printf("LastError?? = %v\n", nc.LastError())
}
```

The above code would result in something like the following below, notice that there is a delay of a few microseconds in getting the error:

```
2015/08/31 19:32:43.010295 Created a message of length: 2097152
2015/08/31 19:32:43.011077 err         = write tcp 127.0.0.1:4222: protocol wrong type for socket
2015/08/31 19:32:43.011091 closed?     = false
2015/08/31 19:32:43.011097 LastError?  = write tcp 127.0.0.1:4222: protocol wrong type for socket
2015/08/31 19:32:43.011194 Disconnected!!!
2015/08/31 19:32:43.011200 closed??    = true
2015/08/31 19:32:43.011205 LastError?? = nats: 'Maximum Payload Violation'
```

On the other hand, the server does give the client the max payload size beforehand on its INFO payload when connecting, so it would be possible to enforce it in the client as well, looking into that too.
Hope this helps!

**Comment by wallyqs** [comment]
Added support to enforcing max payload size on the client level now: https://github.com/nats-io/nats/pull/64

**Comment by jmjava** [comment]
what is the maximum configurable payload size?  We sometimes have to process large messages like 100's of MB... is this possible?

**Comment by derekcollison** [comment]
The server can be configured with a different limit for the message size. Probably would not recommend direct messages that are 100MBs. Can you describe the use case you are trying to solve for?

**Comment by mcqueary** [comment]
Derek's right. Typically with messages that large, we'd recommend "chunking" them (i.e. breaking them into smaller chunks on the producer end and reassembling in sequence on the consumer end), or simply passing URI references to some well-known storage location, vs. attempting to send them as one message. How often is 'sometimes'?  As Derek mentions, more details would be helpful if you don't mind sharing them.

**Comment by jmjava** [comment]
We do calculations that return results that can get quite large.  We wanted
to break up into 2 micro services one to calculate and one to persist
results.

**Comment by xogeny** [comment]
Ironically, my use case was very similar to @jmjava's.  I eventually refactored things to do what @mcqueary suggested (passing around an opaque identifier for the blobs).  Of course, I then needed to have some ubiquitous non-NATS driven system for managing those opaque identifiers.  But it worked out.  My main concern driving that approach was the case of having a blob come along at some point that violated whatever upper bound I established.
